core/src/main/java/org/infinispan/interceptors/locking/ClusteringDependentLogic.java

```diff
@@ -410,14 +443,60 @@ public void init(StateTransferLock stateTransferLock) {
       @Override
       protected void commitSingleEntry(CacheEntry entry, Metadata metadata, FlagAffectedCommand command,
                                        InvocationContext ctx, Flag trackFlag, boolean l1Invalidation) {
+         // Don't allow the CH to change (and state transfer to invalidate entries)
+         // between the ownership check and the commit
          stateTransferLock.acquireSharedTopologyLock();
          try {
-            super.commitSingleEntry(entry, metadata, command, ctx, trackFlag, l1Invalidation);
+            Commit doCommit = shouldCommit(command, ctx, entry.getKey(), entry.isRemoved());
+            if (doCommit.isCommit()) {
+               boolean created = false;
+               boolean removed = false;
+               boolean expired = false;
+               if (doCommit.isLocal()) {
+                  created = entry.isCreated();
+                  removed = entry.isRemoved();
+                  if (removed && entry instanceof MVCCEntry) {
+                     expired = ((MVCCEntry) entry).isExpired();
+                  }
+               }
+
+               InternalCacheEntry previousEntry = dataContainer.peek(entry.getKey());
+               Object previousValue = null;
+               Metadata previousMetadata = null;
+               if (previousEntry != null) {
+                  previousValue = previousEntry.getValue();
+                  previousMetadata = previousEntry.getMetadata();
+               }
+               commitManager.commit(entry, metadata, trackFlag, l1Invalidation);
+               if (doCommit.isLocal()) {
+                  notifyCommitEntry(created, removed, expired, entry, ctx, command, previousValue, previousMetadata);
+               }
+            }
          } finally {
             stateTransferLock.releaseSharedTopologyLock();
          }
       }
 
+      public Commit shouldCommit(FlagAffectedCommand command, InvocationContext ctx, Object key, boolean removed) {
+         // The logic here should be in sync with the logic in DistributionLogic but without the L1 stuff.
+         // Look for comments there
```

Comment:
Oops, outdated, difference is only in `commitSingleEntry`. I should remove one of the methods.